[React Router] Transaction unexpectedly ended early, cancelling spans from componentDidMount()

### Is there an existing issue for this?

- [X] I have checked for existing issues https://github.com/getsentry/sentry-javascript/issues
- [X] I have reviewed the documentation https://docs.sentry.io/
- [X] I am using the latest SDK release https://github.com/getsentry/sentry-javascript/releases

### How do you use Sentry?

Sentry Saas (sentry.io)

### Which package are you using?

@sentry/react

### SDK Version

7.17.3 - but tested on 7.23.0 with same result

### Framework Version

17.0.2 (react-router-dom: 6.4.3)

### Link to Sentry event

https://sentry.io/organizations/testorg-az/performance/application-monitoring-javascript:755fa61c341044a19a38b718efff2b35/?project=5808623&query=sdk.version%3A7.17.3&referrer=performance-transaction-summary&statsPeriod=30d&transaction=%2Fproducts&unselectedSeries=p100()

### Steps to Reproduce

0. Examine our project's SDK configuration [index.js#L60-65](https://github.com/sentry-demos/application-monitoring/blob/master/react/src/index.js#L60-L65) and [index.js#L104](https://github.com/sentry-demos/application-monitoring/blob/master/react/src/index.js#L104)
0.1 Examine code that initiates span that gets cut short before it finishes: [Products.js#L39](https://github.com/sentry-demos/application-monitoring/blob/master/react/src/components/Products.js#L39) - called from `componentDidMount()`
2. Go to https://application-monitoring-react-dot-sales-engineering-sf.appspot.com/products
3. Open devtools
4. (optional) set breakpoint in [idletransaction.tx](https://github.com/getsentry/sentry-javascript/blob/a493aa6a46555b944c8d896a2164bcd8b11caaf5/packages/tracing/src/idletransaction.ts) - will eventually point you to [reactrouterv6.tsx: handleNavigation()](https://github.com/getsentry/sentry-javascript/blob/a493aa6a46555b944c8d896a2164bcd8b11caaf5/packages/react/src/reactrouterv6.tsx#L139) and [_useEffect()](https://github.com/getsentry/sentry-javascript/blob/a493aa6a46555b944c8d896a2164bcd8b11caaf5/packages/react/src/reactrouterv6.tsx#L200)
5. Reload page
6. Inspect console log (Debug level)
7. Get transaction ID from Network tab and plug into search in [/products transaction of this project](https://sentry.io/organizations/testorg-az/performance/summary/?project=5808623&query=&referrer=performance-transaction-summary&statsPeriod=30d&transaction=%2Fproducts&unselectedSeries=p100%28%29)

* Internal ticket: https://getsentry.atlassian.net/browse/FEEDBACK-1679

### Expected Result

Transaction runs its course and does not finish prematurely before `http.client` span (GET https://.../**products**) is complete.

### Actual Result

* Logs show "[Tracing] cancelling span since transaction ended early"
* In Sentry UI the last `http.client` span (GET https://.../**products**) of the transaction is cut short even though corresponding backend transaction is ~ 2 seconds.
<img width="856" alt="Screen Shot 2022-12-07 at 9 10 28 PM" src="https://user-images.githubusercontent.com/490201/206523672-b2321996-e189-407a-9e3b-77be672dae7f.png">

Both of those give me an error because `env-config/local.env` is not checked into the repository.
```
[ERROR] Missing file env-config/local.env or invalid environment 'local'.
```

I've tried creating `env-config/local.env` but what environment variables should I set? I've tried setting the obvious ones and then I get I get [the above error](https://github.com/getsentry/sentry-javascript/issues/6470#issuecomment-1353799874).

--- a/packages/react/src/reactrouterv6.tsx
+++ b/packages/react/src/reactrouterv6.tsx
@@ -170,7 +170,7 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
     return Routes;
   }
 
-  let previousLocation: Location | null = null;
+  let isMountRenderPass: boolean = true;
 
   const SentryRoutes: React.FC<P> = (props: P) => {
     const location = _useLocation();
@@ -180,13 +180,12 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
       () => {
         const routes = _createRoutesFromChildren(props.children) as RouteObject[];
 
-        if (previousLocation === null) {
+        if (isMountRenderPass) {
           updatePageloadTransaction(location, routes);
+          isMountRenderPass = false;
         } else {
           handleNavigation(location, routes, navigationType);
         }
-
-        previousLocation = location;
       },
       // `props.children` is purpusely not included in the dependency array, because we do not want to re-run this effect
       // when the children change. We only want to start transactions when the location or navigation type change.
@@ -215,7 +214,7 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
     return origUseRoutes;
   }
 
-  let previousLocation: Location | null = null;
+  let isMountRenderPass: boolean = true;
 
   // eslint-disable-next-line react/display-name
   return (routes: RouteObject[], locationArg?: Partial<Location> | string): React.ReactElement | null => {
@@ -235,13 +234,12 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
         const normalizedLocation =
           typeof stableLocationParam === 'string' ? { pathname: stableLocationParam } : stableLocationParam;
 
-        if (previousLocation === null) {
+        if (isMountRenderPass) {
           updatePageloadTransaction(normalizedLocation, routes);
+          isMountRenderPass = false;
         } else {
           handleNavigation(normalizedLocation, routes, navigationType);
         }
-
-        previousLocation = normalizedLocation;
       }, [navigationType, stableLocationParam]);
 
       return Routes;

--- a/packages/react/src/reactrouterv6.tsx
+++ b/packages/react/src/reactrouterv6.tsx
@@ -170,35 +170,28 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
     return Routes;
   }
 
-  let routes: RouteObject[];
-  let isFirstPageloadUpdateUseEffectCall = true;
-  let isFirstNavigationUseEffectCall = true;
+  let previousLocation: Location | null = null;
 
   const SentryRoutes: React.FC<P> = (props: P) => {
     const location = _useLocation();
     const navigationType = _useNavigationType();
 
-    _useEffect(() => {
-      // Performance concern:
-      // This is repeated when <Routes /> is rendered.
-      routes = _createRoutesFromChildren(props.children) as RouteObject[];
-    }, [props.children]);
+    _useEffect(
+      () => {
+        const routes = _createRoutesFromChildren(props.children) as RouteObject[];
 
-    _useEffect(() => {
-      if (isFirstPageloadUpdateUseEffectCall) {
-        isFirstPageloadUpdateUseEffectCall = false;
-        routes = _createRoutesFromChildren(props.children) as RouteObject[];
-        updatePageloadTransaction(location, routes);
-      }
-    }, [props.children, location]);
+        if (previousLocation === null) {
+          updatePageloadTransaction(location, routes);
+        } else {
+          handleNavigation(location, routes, navigationType);
+        }
 
-    _useEffect(() => {
-      if (isFirstNavigationUseEffectCall) {
-        isFirstNavigationUseEffectCall = false;
-      } else {
-        handleNavigation(location, routes, navigationType);
-      }
-    }, [location, navigationType]);
+        previousLocation = location;
+      },
+      // `props.children` is purpusely not included in the dependency array, because we do not want to re-run this effect
+      // when the children change. We only want to start transactions when the location or navigation type change.
+      [location, navigationType],
+    );
 
     // @ts-ignore Setting more specific React Component typing for `R` generic above
     // will break advanced type inference done by react router params
@@ -222,35 +215,34 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
     return origUseRoutes;
   }
 
-  let isFirstPageloadUpdateUseEffectCall = true;
-  let isFirstNavigationUseEffectCall = true;
+  let previousLocation: Location | null = null;
 
   // eslint-disable-next-line react/display-name
-  return (routes: RouteObject[], location?: Partial<Location> | string): React.ReactElement | null => {
-    const SentryRoutes: React.FC<unknown> = (props: unknown) => {
-      const Routes = origUseRoutes(routes, location);
+  return (routes: RouteObject[], locationArg?: Partial<Location> | string): React.ReactElement | null => {
+    const SentryRoutes: React.FC<unknown> = () => {
+      const Routes = origUseRoutes(routes, locationArg);
 
-      // the case where location is a string might be killing performance because we're always creating a new
-      // object(with different identity) that will trigger useEffects below to run again
-      const locationArgObject = typeof location === 'string' ? { pathname: location } : location;
-
-      const locationObject = (locationArgObject as Location) || _useLocation();
+      const location = _useLocation();
       const navigationType = _useNavigationType();
 
-      _useEffect(() => {
-        if (isFirstPageloadUpdateUseEffectCall) {
-          isFirstPageloadUpdateUseEffectCall = false;
-          updatePageloadTransaction(locationObject, routes);
-        }
-      }, [locationObject, routes]);
+      // A value with stable identity to either pick `locationArg` if available or `location` if not
+      const stableLocationParam =
+        typeof locationArg === 'string' || locationArg?.pathname !== undefined
+          ? (locationArg as { pathname: string })
+          : location;
 
       _useEffect(() => {
-        if (isFirstNavigationUseEffectCall) {
-          isFirstNavigationUseEffectCall = false;
+        const normalizedLocation =
+          typeof stableLocationParam === 'string' ? { pathname: stableLocationParam } : stableLocationParam;
+
+        if (previousLocation === null) {
+          updatePageloadTransaction(normalizedLocation, routes);
         } else {
-          handleNavigation(locationObject, routes, navigationType);
+          handleNavigation(normalizedLocation, routes, navigationType);
         }
-      }, [locationObject, routes, navigationType]);
+
+        previousLocation = normalizedLocation;
+      }, [navigationType, stableLocationParam]);
 
       return Routes;
     };

--- a/packages/react/src/reactrouterv6.tsx
+++ b/packages/react/src/reactrouterv6.tsx
@@ -131,17 +131,8 @@ function handleNavigation(
   location: Location,
   routes: RouteObject[],
   navigationType: Action,
-  isBaseLocation: boolean,
   matches?: AgnosticDataRouteMatch,
 ): void {
-  if (isBaseLocation) {
-    if (activeTransaction) {
-      activeTransaction.finish();
-    }
-
-    return;
-  }
-
   const branches = Array.isArray(matches) ? matches : _matchRoutes(routes, location);
 
   if (_startTransactionOnLocationChange && (navigationType === 'PUSH' || navigationType === 'POP') && branches) {
@@ -179,12 +170,12 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
     return Routes;
   }
 
-  let isBaseLocation: boolean = false;
   let routes: RouteObject[];
 
   const SentryRoutes: React.FC<P> = (props: P) => {
     const location = _useLocation();
     const navigationType = _useNavigationType();
+    let isBaseLocation: boolean = false;
 
     _useEffect(() => {
       // Performance concern:
@@ -197,7 +188,9 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
     }, [props.children]);
 
     _useEffect(() => {
-      handleNavigation(location, routes, navigationType, isBaseLocation);
+      if (!isBaseLocation) {
+        handleNavigation(location, routes, navigationType);
+      }
     }, [props.children, location, navigationType, isBaseLocation]);
 
     isBaseLocation = false;
@@ -224,8 +217,6 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
     return origUseRoutes;
   }
 
-  let isBaseLocation: boolean = false;
-
   // eslint-disable-next-line react/display-name
   return (routes: RouteObject[], location?: Partial<Location> | string): React.ReactElement | null => {
     const SentryRoutes: React.FC<unknown> = (props: unknown) => {
@@ -234,6 +225,7 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
       const locationArgObject = typeof location === 'string' ? { pathname: location } : location;
       const locationObject = (locationArgObject as Location) || _useLocation();
       const navigationType = _useNavigationType();
+      let isBaseLocation: boolean = false;
 
       _useEffect(() => {
         isBaseLocation = true;
@@ -242,7 +234,9 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
       }, [props]);
 
       _useEffect(() => {
-        handleNavigation(locationObject, routes, navigationType, isBaseLocation);
+        if (!isBaseLocation) {
+          handleNavigation(locationObject, routes, navigationType);
+        }
       }, [props, locationObject, navigationType, isBaseLocation]);
 
       isBaseLocation = false;
@@ -275,7 +269,7 @@ export function wrapCreateBrowserRouter(createRouterFunction: CreateRouterFuncti
         (state.historyAction === 'PUSH' || state.historyAction === 'POP') &&
         activeTransaction
       ) {
-        handleNavigation(location, routes, state.historyAction, false);
+        handleNavigation(location, routes, state.historyAction);
       }
     });
 

--- a/packages/react/src/reactrouterv6.tsx
+++ b/packages/react/src/reactrouterv6.tsx
@@ -185,20 +185,20 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
     }, [props.children]);
 
     _useEffect(() => {
-      if (isFirstNavigationUseEffectCall) {
-        isFirstNavigationUseEffectCall = false;
-      } else {
-        handleNavigation(location, routes, navigationType);
-      }
-    }, [location, navigationType]);
-
-    _useEffect(() => {
       if (isFirstPageloadUpdateUseEffectCall) {
         isFirstPageloadUpdateUseEffectCall = false;
         routes = _createRoutesFromChildren(props.children) as RouteObject[];
         updatePageloadTransaction(location, routes);
       }
     }, [props.children, location]);
+
+    _useEffect(() => {
+      if (isFirstNavigationUseEffectCall) {
+        isFirstNavigationUseEffectCall = false;
+      } else {
+        handleNavigation(location, routes, navigationType);
+      }
+    }, [location, navigationType]);
 
     // @ts-ignore Setting more specific React Component typing for `R` generic above
     // will break advanced type inference done by react router params

--- a/packages/react/src/reactrouterv6.tsx
+++ b/packages/react/src/reactrouterv6.tsx
@@ -171,29 +171,35 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
   }
 
   let routes: RouteObject[];
+  let isFirstPageloadUpdateUseEffectCall = true;
+  let isFirstNavigationUseEffectCall = true;
 
   const SentryRoutes: React.FC<P> = (props: P) => {
     const location = _useLocation();
     const navigationType = _useNavigationType();
-    let isBaseLocation: boolean = false;
 
     _useEffect(() => {
       // Performance concern:
       // This is repeated when <Routes /> is rendered.
       routes = _createRoutesFromChildren(props.children) as RouteObject[];
-      isBaseLocation = true;
-
-      updatePageloadTransaction(location, routes);
-      // eslint-disable-next-line react-hooks/exhaustive-deps
     }, [props.children]);
 
     _useEffect(() => {
-      if (!isBaseLocation) {
+      if (isFirstNavigationUseEffectCall) {
+        isFirstNavigationUseEffectCall = false;
+      } else {
         handleNavigation(location, routes, navigationType);
       }
-    }, [props.children, location, navigationType, isBaseLocation]);
+    }, [location, navigationType]);
 
-    isBaseLocation = false;
+    _useEffect(() => {
+      if (isFirstPageloadUpdateUseEffectCall) {
+        routes = _createRoutesFromChildren(props.children) as RouteObject[];
+        updatePageloadTransaction(location, routes);
+      } else {
+        isFirstPageloadUpdateUseEffectCall = false;
+      }
+    }, [props.children, location]);
 
     // @ts-ignore Setting more specific React Component typing for `R` generic above
     // will break advanced type inference done by react router params
@@ -217,29 +223,36 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
     return origUseRoutes;
   }
 
+  let isFirstPageloadUpdateUseEffectCall = true;
+  let isFirstNavigationUseEffectCall = true;
+
   // eslint-disable-next-line react/display-name
   return (routes: RouteObject[], location?: Partial<Location> | string): React.ReactElement | null => {
     const SentryRoutes: React.FC<unknown> = (props: unknown) => {
       const Routes = origUseRoutes(routes, location);
 
+      // the case where location is a string might be killing performance because we're always creating a new
+      // object(with different identity) that will trigger useEffects below to run again
       const locationArgObject = typeof location === 'string' ? { pathname: location } : location;
+
       const locationObject = (locationArgObject as Location) || _useLocation();
       const navigationType = _useNavigationType();
-      let isBaseLocation: boolean = false;
 
       _useEffect(() => {
-        isBaseLocation = true;
-
-        updatePageloadTransaction(locationObject, routes);
-      }, [props]);
-
-      _useEffect(() => {
-        if (!isBaseLocation) {
-          handleNavigation(locationObject, routes, navigationType);
+        if (isFirstPageloadUpdateUseEffectCall) {
+          updatePageloadTransaction(locationObject, routes);
+        } else {
+          isFirstPageloadUpdateUseEffectCall = false;
         }
-      }, [props, locationObject, navigationType, isBaseLocation]);
+      }, [locationObject, routes]);
 
-      isBaseLocation = false;
+      _useEffect(() => {
+        if (isFirstNavigationUseEffectCall) {
+          handleNavigation(locationObject, routes, navigationType);
+        } else {
+          isFirstNavigationUseEffectCall = false;
+        }
+      }, [locationObject, routes, navigationType]);
 
       return Routes;
     };

--- a/packages/react/src/reactrouterv6.tsx
+++ b/packages/react/src/reactrouterv6.tsx
@@ -194,10 +194,9 @@ export function withSentryReactRouterV6Routing<P extends Record<string, any>, R 
 
     _useEffect(() => {
       if (isFirstPageloadUpdateUseEffectCall) {
+        isFirstPageloadUpdateUseEffectCall = false;
         routes = _createRoutesFromChildren(props.children) as RouteObject[];
         updatePageloadTransaction(location, routes);
-      } else {
-        isFirstPageloadUpdateUseEffectCall = false;
       }
     }, [props.children, location]);
 
@@ -240,17 +239,16 @@ export function wrapUseRoutes(origUseRoutes: UseRoutes): UseRoutes {
 
       _useEffect(() => {
         if (isFirstPageloadUpdateUseEffectCall) {
-          updatePageloadTransaction(locationObject, routes);
-        } else {
           isFirstPageloadUpdateUseEffectCall = false;
+          updatePageloadTransaction(locationObject, routes);
         }
       }, [locationObject, routes]);
 
       _useEffect(() => {
         if (isFirstNavigationUseEffectCall) {
-          handleNavigation(locationObject, routes, navigationType);
-        } else {
           isFirstNavigationUseEffectCall = false;
+        } else {
+          handleNavigation(locationObject, routes, navigationType);
         }
       }, [locationObject, routes, navigationType]);
 

